src/nbl/video/CVulkanPhysicalDevice.h

```diff
@@ -101,72 +100,163 @@ class CVulkanPhysicalDevice final : public IPhysicalDevice
         }
     }
             
-    inline VkPhysicalDevice getInternalObject() const { return m_vkphysdev; }
+    inline VkPhysicalDevice getInternalObject() const { return m_vkPhysicalDevice; }
             
     E_API_TYPE getAPIType() const override { return EAT_VULKAN; }
 
-    std::vector<ISurface::SFormat> getAvailableFormatsForSurface(const ISurface* surface) const override
+    // Todo(achal)
+    IDebugCallback* getDebugCallback() override { return nullptr; }
+
+    void getAvailableFormatsForSurface(const ISurface* surface, uint32_t& formatCount,
+        ISurface::SFormat* formats) const override
     {
-        const ISurfaceVK* vk_surface = static_cast<const ISurfaceVK*>(surface); // Todo(achal): This is problematic, if passed `surface` isn't a vulkan surface
+        constexpr uint32_t MAX_SURFACE_FORMAT_COUNT = 1000u;
+
+        if (surface->getAPIType() != EAT_VULKAN)
+            return;
 
-        uint32_t formatCount = 0u;
-        vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkphysdev, vk_surface->m_surface, &formatCount, nullptr);
-        std::vector<VkSurfaceFormatKHR> formats(formatCount);
-        vkGetPhysicalDeviceSurfaceFormatsKHR(m_vkphysdev, vk_surface->m_surface, &formatCount, formats.data());
+        // Todo(achal): not sure yet, how would I handle multiple platforms without making
+        // this function templated
+        VkSurfaceKHR vk_surface = static_cast<const CSurfaceVulkanWin32*>(surface)->getInternalObject();
```

Comment:
it would be best if the available formats function was a swapchain method taking a physical device (swapchain self reports)

this way we can have "foreign" swapchains as part of application integrations